File: for-jython/ubmpsupport.py
# unit ubmpsupport

from conversion_common import *
"""
import umain
"""
import delphi_compatability

def resizeBitmap(bitmap, newSize):
    if bitmap == None:
        return
    try:
        bitmap.Width = newSize.X
        bitmap.Height = newSize.Y
    except:
        bitmap.Width = 1
        bitmap.Height = 1

def fillBitmap(gc, bitmap, color):
    if bitmap == None:
        return
    bitmap.Fill(gc, color)

def copyBitmapToCanvasWithGlobalPalette(gc, bitmap, aCanvas, aRect):
    aCanvas.draw_drawable(gc, bitmap.pixmap, 0, 0, aRect.left, aRect.top, aRect.right - aRect.left, aRect.bottom - aRect.top)
    return

    ## UNUSED
    oldPalette = HPALETTE()
    wholeRect = TRect()
    needPalette = False
    
    oldPalette = 0
    needPalette = screenColorModeIs256ColorsOrLess()
    if needPalette:
        oldPalette = UNRESOLVED.selectPalette(aCanvas.Handle, umain.MainForm.paletteImage.Picture.Bitmap.Palette, False)
        UNRESOLVED.realizePalette(aCanvas.Handle)
    wholeRect = Rect(0, 0, bitmap.Width, bitmap.Height)
    if (aRect.Left != 0) or (aRect.Right != 0) or (aRect.Top != 0) or (aRect.Bottom != 0):
        aCanvas.CopyRect(aRect, bitmap.Canvas, wholeRect)
    else:
        aCanvas.CopyRect(wholeRect, bitmap.Canvas, wholeRect)
    if needPalette:
        UNRESOLVED.selectPalette(aCanvas.Handle, oldPalette, True)
        UNRESOLVED.realizePalette(aCanvas.Handle)

def screenColorModeIs256ColorsOrLess():
    # PDF PORT __ PUT IN DEFAULT FOR TESTING __ PROBABL OK THOUGH
    return False
    result = False
    screenDC = UNRESOLVED.GetDC(0)
    try:
        screenColorBits = (UNRESOLVED.GetDeviceCaps(screenDC, delphi_compatability.BITSPIXEL) * UNRESOLVED.GetDeviceCaps(screenDC, delphi_compatability.PLANES))
    finally:
        UNRESOLVED.ReleaseDC(0, screenDC)
    result = (screenColorBits == 1) or (screenColorBits == 4) or (screenColorBits == 8)
    return result

def setPixelFormatBasedOnScreenForBitmap(bitmap):
    if bitmap == None:
        return
    if screenColorModeIs256ColorsOrLess():
        bitmap.PixelFormat = delphi_compatability.TPixelFormat.pf8bit
        bitmap.Palette = UNRESOLVED.CopyPalette(umain.MainForm.paletteImage.Picture.Bitmap.Palette)

File: converted_source/ubmpsupport.pas
unit ubmpsupport;

interface

uses Windows, Graphics, Classes;

procedure resizeBitmap(bitmap: TBitmap; newSize: TPoint);
procedure fillBitmap(bitmap: TBitmap; color: TColor);
procedure copyBitmapToCanvasWithGlobalPalette(bitmap: TBitmap; aCanvas: TCanvas; aRect: TRect);
function screenColorModeIs256ColorsOrLess: boolean;
procedure setPixelFormatBasedOnScreenForBitmap(bitmap: TBitmap);

implementation

uses umain;

procedure resizeBitmap(bitmap: TBitmap; newSize: TPoint);
  begin
  if bitmap = nil then exit;
  with bitmap do
    try
      width := newSize.x;
      height := newSize.y;
    except
      width := 1;
      height := 1;
    end;
  end;

procedure fillBitmap(bitmap: TBitmap; color: TColor);
  begin
  if bitmap = nil then exit;
  with bitmap.canvas do
    begin
    brush.color := color;
    fillRect(rect(0, 0, bitmap.width, bitmap.height));
    end;
  end;

procedure copyBitmapToCanvasWithGlobalPalette(bitmap: TBitmap; aCanvas: TCanvas; aRect: TRect);
  var
    oldPalette: HPALETTE;
    wholeRect: TRect;
    needPalette: boolean;
  begin
  oldPalette := 0;
  needPalette := screenColorModeIs256ColorsOrLess;
  if needPalette then
    begin
    oldPalette := selectPalette(aCanvas.handle, MainForm.paletteImage.picture.bitmap.palette, false);
    realizePalette(aCanvas.handle);
    end;
  wholeRect := Rect(0, 0, bitmap.width, bitmap.height);
  with aRect do if (left <> 0) or (right <> 0) or (top <> 0) or (bottom <> 0) then
    aCanvas.copyRect(aRect, bitmap.canvas, wholeRect)
  else
    aCanvas.copyRect(wholeRect, bitmap.canvas, wholeRect);
  if needPalette then
    begin
    selectPalette(aCanvas.handle, oldPalette, true);
    realizePalette(aCanvas.handle);
    end;
  end;

function screenColorModeIs256ColorsOrLess: boolean;
  var
    screenDC: HDC;
    screenColorBits: integer;
  begin
  result := false;
  screenDC := GetDC(0);
  try
    screenColorBits := (GetDeviceCaps(screenDC, BITSPIXEL) * GetDeviceCaps(screenDC, PLANES));
  finally
    ReleaseDC(0, screenDC);
  end;
  result := (screenColorBits = 1) or (screenColorBits = 4) or (screenColorBits = 8);
  end;

procedure setPixelFormatBasedOnScreenForBitmap(bitmap: TBitmap);
  begin
  if bitmap = nil then exit;
  if screenColorModeIs256ColorsOrLess then
    begin
    bitmap.pixelFormat := pf8bit;
    bitmap.palette := CopyPalette(MainForm.paletteImage.picture.bitmap.palette);
    end;
  end;

end.
 